source/auth/auth.rst

```diff
@@ -250,6 +250,48 @@ or "SCRAM-SHA-256").
 
+Speculative Authentication via Handshake
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+:since: 4.4
+
+The ``isMaster`` handshake supports a new argument, ``speculativeSaslStart``,
+provided as a BSON object. Clients specifying this argument to ``isMaster`` will
+speculatively include the first ``saslStart`` command of an authentication handshake.
+This command may be provided to the server in parallel with any standard request for
+supported authentication mechanisms. This would permit clients to merge their
+``saslStart`` message with their ``isMaster`` request, and receive the ``saslStart``
+reply with the ``isMaster`` reply.
+
+If the speculatively issued command succeeds, the client should proceed with the next
+step of the exchange. If the command fails, the client should determine whether the
+``isMaster`` response contains enough information to select another mechanism.
+
+Older servers will ignore this argument. New servers will participate in the standard
+authentication conversation if this argument is missing.
+
+An example conversation using ``speculativeSaslStart`` is as follows:
+
+| C: :javascript:`{ isMaster: 1, speculativeSaslStart: { mechanism: "SCRAM-SHA-256", db: "admin", payload: BinData(0, "biwsbj11c2VyLHI9ZnlrbytkMmxiYkZnT05Sdjlxa3hkYXdM"), options: { skipEmptyExchange: true }}, ... }`
+| S: :javascript:`{ isMaster: true, speculativeSaslStart: { conversationId: 1, payload: BinData(0,"cj1meWtvK2QybGJiRmdPTlJ2OXFreGRhd0xIbytWZ2s3cXZVT0tVd3VXTElXZzRsLzlTcmFHTUhFRSxzPXJROVpZM01udEJldVAzRTFURFZDNHc9PSxpPTEwMDAw"), done: false }, ... }`
+| C: :javascript:`{saslContinue: 1, conversationId: 1, payload: BinData(0, "Yz1iaXdzLHI9ZnlrbytkMmxiYkZnT05Sdjlxa3hkYXdMSG8rVmdrN3F2VU9LVXd1V0xJV2c0bC85U3JhR01IRUUscD1NQzJUOEJ2Ym1XUmNrRHc4b1dsNUlWZ2h3Q1k9")}`
+| S: :javascript:`{conversationId: 1, payload: BinData(0,"dj1VTVdlSTI1SkQxeU5ZWlJNcFo0Vkh2aFo5ZTA9"), done: true, ok: 1}`
+  
+
+Alternatively, the ``isMaster`` handshake supports another new argument,
+``speculativeAuthenticate``, provided as a BSON object. This argument provides an
+instance of the ``authenticate`` command with the ``MONGODB-X509`` mechanism.
+Servers accepting the request will respond with an ``isMaster`` reply containing the results
+in a field named ``speculativeAuthenticate``. Concurrent use of ``speculativeSaslStart``
+and ``speculativeAuthenticate`` will result in ``isMaster`` returning an error.  In this
+mode, authentication can occur in zero round trips and will be completed once ``isMaster``
+returns.
+
+An example conversation using ``speculativeAuthenticate`` is as follows:
+
+| C: :javascript:`{ isMaster: 1, speculativeAuthenticate: { mechanism: "MONGODB-X509", db: "$external" }, ... }`
+| S: :javascript:`{ ismaster: true, speculativeAuthenticate: { dbname: "$external", user : "CN=client,OU=KernelUser,O=MongoDB,L=New York City,ST=New York,C=US" }, ok: 1, ... }`
+
 --------------------------------
 Supported Authentication Methods
```

Comment:
(Commenting here because I can't comment below.) Can we add something to the "Test Plan" section about this feature?

Reply:
I have added a section to the Test Plan for speculative authentication.

Reply:
I no longer see the test plan in the latest commit.  So I'm not sure.

I think my point was that, given that we don't have any events that are published that prove that the driver is actually doing speculative authentication (unlike with CRUD, where we can use command monitoring to assert the internal behavior of the driver), the best we can do is to just integration test all the authentication methods against server versions which support speculative authentication and server versions that don't.  Drivers generally already have these tests (at least for SCRAM... There are at least some drivers that don't for X509 auth).

Some drivers are designed in a way where this behavior is unit-testable, but I'm not sure we should make that a requirement in the spec (like we do implicitly for SDAM, for example).

Reply:
Ok, I agree with this. I think it is best to remove the test plan section, since it would be redundant to state that drivers should run the integration tests that already exist.

Reply:
@ShaneHarvey what do you think?